QueryOptions parameters (GjkOptions)

ThierryBerger opened the pull request:
Following [discussion on discord](https://discordapp.com/channels/507548572338880513/748627261384556715/1316845556546539542), a shape cast of a ball sometimes fails when it should not.

- Related to https://github.com/dimforge/parry/issues/180
- Would fix:
  - https://github.com/dimforge/parry/issues/70
  - https://github.com/dimforge/parry/issues/317
  - https://github.com/dimforge/parry/issues/193

## Current status

- [x] Exposes gjk parameters through `GjkOptions`, this is very breaking as it impacts `PointQuery` trait.
  - [x] `DefaultQueryDispatcher` stores relevant options to pass to their algorithms.
    - [ ] store an option dispatcher inside the `DefaultQueryDispatcher` to handle user-defined algorithms (and more elegantly our convex polygons).
      - [ ] Fix `intersection_test_point_query_ball` (https://github.com/dimforge/parry/pull/298/files#r2230938921)
- [x] Use a `QueryOoptionsNotUsed` in algorithms to make it easier to understand which paths need an option and which don´t. Ideally an associated trait would be used but it makes it difficult to mix different shapes...
- :thinking: Maybe we should consider making new methods to pass parameters, like `contains_local_point_with_option` ? to make the API more straightforward by default, but I'm not a huge fan of duplicated entries
- :thinking: Can we improve the type safety ? Currently an incorrect parameter leads to using default and logging a warning, I think it's good enough :tm: but not amazing 

Before this PR:

![Screenshot failed](https://github.com/user-attachments/assets/824dfbfc-b43b-4917-8785-0aa5b24433fd)



After this PR, with customizable Gjk epsilon changing each frame:

[Screencast from 07-24-2025 04:49:01 PM.webm](https://github.com/user-attachments/assets/5fb6c8fa-d5e0-48d4-8f14-015b271abccc)



I added an example for compound shape point query (using convex polygon, to test dispatching gjk options), and the result was showing false negatives (without custom gjk options ; multiplying the epsilon by 100 fixes it.):

[Screencast from 07-16-2025 04:09:26 PM.webm](https://github.com/user-attachments/assets/847d1aef-ba5a-4bdb-a860-70f409e2f2d4)

ThierryBerger committed:
heavy wip ; I could reproduce a wrong shapecast :thinking:

ThierryBerger committed:
better example with all cases shown at once

ThierryBerger committed:
up epsilon for gjk

ThierryBerger committed:
add changelog

ThierryBerger committed:
removed incorrect lint

ThierryBerger committed:
add example to cargo toml + rename with 2d postfix

Vrixyz commented on src/query/gjk/gjk.rs:
```diff
@@ -35,7 +35,7 @@ pub enum GJKResult {
 /// The absolute tolerance used by the GJK algorithm.
 pub fn eps_tol() -> Real {
     let _eps = crate::math::DEFAULT_EPSILON;
-    _eps * 10.0
+    _eps * 10_000.0
 }
```
I'm not a huge fan of increasing epsilons arbitrarily, but that improves the example added in this PR.

I think exposing this value would be more interesting, and I'd like to have a test stressing false positive, where a high epsilon results in an incorrect result :thinking:

Vrixyz commented on crates/parry2d/examples/debug_shape_cast2d.rs:
```diff
@@ -0,0 +1,185 @@
+            Point2::new(screen_width() / 2.0, screen_height() / 2.0) + Vector2::new(-500.0, -200.0);
+
+        let to_cast_against = ConvexPolygon::from_convex_polyline(
+            [
+                [600.0, 288.0].into(),
+                [576.0, 312.0].into(),
+                [552.0, 288.0].into(),
+                [576.0, 264.0].into(),
+            ]
+            .into(),
```
check if that works on a centered shape.

Vrixyz replied:
It doesn't work better with a centered shape. I updated the code to have the shape centered.

ThierryBerger committed:
center shape to raycast

Vrixyz commented on crates/parry2d/examples/debug_shape_cast2d.rs:
```diff
@@ -0,0 +1,188 @@
+
+        let to_cast_against = ConvexPolygon::from_convex_polyline(
+            [
+                [-24.0, 0.0].into(),
+                [0.0, 24.0].into(),
+                [24.0, 0.0].into(),
+                [0.0, -24.0].into(),
+            ]
```
todo: make a real unit test for that

ThierryBerger committed:
add unit test

ThierryBerger committed:
add ability to customize gjk at the dispatcher level (still needs to fix more path to eps_tol(), but serves as discussion starter

Vrixyz commented on src/query/default_query_dispatcher.rs:
```diff
@@ -14,7 +15,18 @@ use crate::shape::{HalfSpace, Segment, Shape, ShapeType};
-pub struct DefaultQueryDispatcher;
+pub struct DefaultQueryDispatcher {
+    /// The absolute tolerance used by the GJK algorithm.
+    pub gjk_espilon_tolerance: Real,
```
Create a `GjkOptions` struct ; implements default ; add `nb_max_iterations` ; check article gjk to add doc on that.

ThierryBerger committed:
add dedicated gjk options struct and passed in most convenient places

ThierryBerger committed:
wip: using point query options as dyn ; dispatching them depending on shapes knowing their underlying algorithm.

ThierryBerger committed:
add an example to test out API for compound shape options

It's not straightforward to do, that will likely need a solution similar to a dispatcher

ThierryBerger committed:
Merge branch 'master' into incorrect-shapecast

ThierryBerger committed:
QueryOptionsDispatcherMap makes it possible to forward options to a compound shape.

ThierryBerger committed:
code cleanup

Vrixyz commented on CHANGELOG.md:
```diff
@@ -133,6 +133,8 @@
 - `point_cloud_bounding_sphere` and `point_cloud_bounding_sphere_with_center` now returns a `BoundingSphere`.
 - Removed `IntersectionCompositeShapeShapeBestFirstVisitor` (which had been deprecated for a while):
   use `IntersectionCompositeShapeShapeVisitor` instead.
+- Epsilon from `parry::query::gjk::eps_tol` is now `10e-2` (from `10e-5`).
+  - This fixes cases of incorrectly failing shapecasts.
```
This changelog is now incorrect

ThierryBerger committed:
more docs + fix default gjk epsilon

ThierryBerger committed:
fix bench + fix default epsilon used in QueryOptions default

Vrixyz commented:
- Current CI failing due to https://github.com/dimforge/parry/pull/364

ThierryBerger committed:
update changelog

Vrixyz commented on CHANGELOG.md:
```diff
@@ -1,3 +1,10 @@
+### Unreleased
```
```suggestion
# Unreleased
```

ThierryBerger committed:
more consistent headings

Vrixyz commented on src/query/point/point_composite_shape/query_options_dispatcher.rs:
```diff
@@ -0,0 +1,96 @@
+            });
+        QueryOptionsDispatcherMap {
+            options_for_shape: [
+                (TypeId::of::<Compound>(), self_ref),
+                #[cfg(feature = "dim2")]
+                (
+                    TypeId::of::<crate::shape::ConvexPolygon>(),
+                    gjk_options.clone(),
+                ),
+                #[cfg(feature = "dim3")]
+                (TypeId::of::<crate::shape::ConvexPolyhedron>(), gjk_options),
```
Do roundshapes and other container shapes need to be specific ? like `TypeId::of::<RolunShape<crate::shape::ConvexPolyhedron>>()` ?

Vrixyz replied:
roundshapes use support point algorithm which is gjk options.

Vrixyz commented on src/query/ray/ray_round_shape.rs:
```diff
@@ -16,6 +16,8 @@ impl<S: SupportMap> RayCast for RoundShape<S> {
             max_time_of_impact,
             solid,
+            // TODO: be able to pass custom option.
+            &GjkOptions::default(),
```
See previous comment

ThierryBerger committed:
Merge branch 'master' into incorrect-shapecast

ThierryBerger committed:
wip ray support

ThierryBerger committed:
Made more clear when query options are needed or not, adapted example to change epsilon each frame

Vrixyz commented on src/query/default_query_dispatcher.rs:
```diff
@@ -42,11 +46,23 @@ impl QueryDispatcher for DefaultQueryDispatcher {
+                pos12,
+                shape1,
+                b2,
+                // FIXME: This needs a query option dispatcher, because a user custom shape against a ball may need a different option.
+                // Currently, some paths lead to an unused option.
+                // (Gjk is used for intersections with convex shapes and cylinder)
+                &self.gjk_options,
```
This is an important point to tackle.
